Add procedure to upgrade Foreman on Deb (#3225)
theforeman/foreman-documentation: guides/common/modules/proc_upgrading-a-connected-project-server.adoc

# foreman-rake katello:upgrade_check
----
endif::[]
ifdef::foreman-deb[]
. Update repositories in `/etc/apt/sources.list.d/foreman.list` on your {ProjectServer} to consume content for {Project} {ProjectVersion}:
+
[options="nowrap" subs="attributes"]
----
deb http://deb.theforeman.org/ _My_Distribution_Codename_ {ProjectVersion}
deb http://deb.theforeman.org/ plugins {ProjectVersion}
----
+
// List of supported OS for foreman-deb has to match "guides/common/modules/proc_configuring-repositories-foreman-deb.adoc"

> Replace _My_Distribution_Codename_ with `bullseye` if you run {ProjectServer} on Debian 11, `jammy` for Ubuntu 22.04, and `focal` for Ubuntu 20.04.

Reviewer: I'd like to see all the items in the list use the same structure (parallelism). It tripped me up the first time I read this. How about this:

```suggestion
Replace _My_Distribution_Code_Name_ with the proper distribution code name based on where your run {ProjectServer}:

* `bullseye` for Debian 11
* `jammy` for Ubuntu 22.04
* `focal` for Ubuntu 20.04
```

Not sure if you really need the list.

Also, "code name" based on [Merriam Webster](https://www.merriam-webster.com/dictionary/code%20name).

Author: Applied.

. Update to {Project} {ProjectVersion}:
+
[options="nowrap" subs="attributes"]
----
# apt-get update
# apt-get --only-upgrade install ruby\* foreman\*
----

> . Migrate the database on {ProjectServer}:

Author: Should I drop "on {ProjectServer}" everywhere because it's obvious?

Reviewer: All the steps should be performed on {ProjectServer}, is that correct? You could just write it once at the very beginning:

```
.Upgrade {ProjectServer}
On your {ProjectServer}:

<step 1>
```

Author: Dropped "on {ProjectServer}"; the procedure is named ".Upgrade {ProjectServer}".

+
[options="nowrap" subs="attributes"]
----
# foreman-rake db:migrate
# foreman-rake db:seed
----
. Clear the cache and existing database sessions on {ProjectServer}:
+
[options="nowrap" subs="attributes"]
----
# foreman-rake tmp:cache:clear
# foreman-rake db:sessions:clear
----
. Optional: Reclaim space on your PostgreSQL database on {ProjectServer}:
+
[options="nowrap" subs="attributes"]
----
# su - postgres -c "vacuumdb --full --dbname=foreman"
----

> . If you have used `{foreman-installer}` to set up your existing {ProjectServer}, {Team} recommends running it again after upgrading.
> * `{foreman-installer}` can modify config files so this may overwrite your custom changes.

Reviewer: ```suggestion
. If you have used `{foreman-installer}` to set up your existing {ProjectServer}, {Team} recommends running it again after upgrading.
However, `{foreman-installer}` can modify config files so this might overwrite your custom changes.
```

Technically, I think line 119 is more tightly related to line 118 than 120. Which is why I'm suggesting this.

You could even consider using an admonition. I know that they are frowned upon in this team :) but they do exist for a reason. Your call.

```suggestion
. If you have used `{foreman-installer}` to set up your existing {ProjectServer}, {Team} recommends running it again after upgrading.
+
[IMPORTANT]
====
`{foreman-installer}` can modify config files so this might overwrite your custom changes.
====
```

Author: Applied your first suggestion without admonition.

> You can run the installer in noop mode so you can see what would be changed:

Reviewer: ```suggestion
You can run the installer in noop mode to see what would be changed:
```

Author: Applied.

+
[options="nowrap" subs="+quotes,verbatim,attributes"]
----
# {foreman-installer} --noop --verbose
----
+

> You may see errors such as `/Stage[main]/Foreman_proxy::Register/Foreman_smartproxy[{foreman-example-com}]: Could not evaluate: Connection refused - connect(2)` due to apache2 service being stopped.
> These can be safely ignored.

Reviewer: Just a suggestion to make this easier to parse.

```suggestion
You might see errors such as the following:
+
----
/Stage[main]/Foreman_proxy::Register/Foreman_smartproxy[{foreman-example-com}]: Could not evaluate: Connection refused - connect(2)
 ----
 +
These errors mean that the `apache2` service is being stopped.
You can safely ignore them.
```

Author: Applied.

> * To apply these changes, run the installer again without options:

Reviewer: I don't think "applying changes" is an apt enough description for what we're asking users to do. How about this?

```suggestion
* After reviewing the changes the installer will make, run it again:
```

Author: Applied.

# foreman-rake katello:upgrade_check
----
endif::[]
ifdef::foreman-deb[]
. Update repositories in `/etc/apt/sources.list.d/foreman.list` to consume content for {Project} {ProjectVersion}:
+
[options="nowrap" subs="attributes"]
----
deb http://deb.theforeman.org/ _My_Distribution_Code_Name_ {ProjectVersion}
deb http://deb.theforeman.org/ plugins {ProjectVersion}
----
+
// List of supported OS for foreman-deb has to match "guides/common/modules/proc_configuring-repositories-foreman-deb.adoc"
Replace _My_Distribution_Code_Name_ with the proper distribution code name based on the operating system of your {ProjectServer}:
+
* `bullseye` for Debian 11
* `jammy` for Ubuntu 22.04
* `focal` for Ubuntu 20.04
. Update to {Project} {ProjectVersion}:
+
[options="nowrap" subs="attributes"]
----
# apt-get update
# apt-get --only-upgrade install ruby\* foreman\*
----

> . Migrate the database:

Reviewer 2: I would prefer to take this opportunity and drop all non-installer setups from the docs.
Then the upgrade becomes:
1. stop services
2. apt update/apt upgrade
3. foreman-installer

Reviewer 2: (we do so in EL-docs already.

# foreman-rake katello:upgrade_check
----
endif::[]
ifdef::foreman-deb[]
. Update repositories in `/etc/apt/sources.list.d/foreman.list` to consume content for {Project} {ProjectVersion}:
+
[options="nowrap" subs="+quotes,verbatim,attributes"]
----
deb http://deb.theforeman.org/ _My_Distribution_Code_Name_ {ProjectVersion}
deb http://deb.theforeman.org/ plugins {ProjectVersion}
----
+
// List of supported OS for foreman-deb has to match "guides/common/modules/proc_configuring-repositories-foreman-deb.adoc"
Replace _My_Distribution_Code_Name_ with the proper distribution code name based on the operating system of your {ProjectServer}:
+
* `bullseye` for Debian 11
* `jammy` for Ubuntu 22.04
* `focal` for Ubuntu 20.04
. Stop all {Project} services:
+
[options="nowrap" subs="+quotes,verbatim,attributes"]
----
# {foreman-maintain} service stop
----
. Update to {Project} {ProjectVersion}:
+
[options="nowrap" subs="attributes"]
----
# apt-get update
# apt-get upgrade

> ----

Reviewer 2: The flow here is a bit weird:
* change repo config (apt)
* stop services (foreman)
* update packages (apt)

While technically correct (changing repo config doesn't further alter the system), I'd still go with:
* stop services (foreman)
* change repo config (apt)
* update packages (apt)

As it creates less context switches.

Author: Makes sense, fixed.

. Restart all {Project} services:
+
[options="nowrap" subs="+quotes,verbatim,attributes"]
----

> # {foreman-maintain} service restart

Author: Removed this code block; added a command to stop all services before performing the package upgrade.
